File: python/01696_jump_game_VI.py
# Time complexity - O (N)
# Space complexity - O (N)
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]

        dq = deque()
        dq.append(0)

        for i in range(1, n):
            dp[i] = nums[i] + dp[dq[0]]

            while dq and dq[0] <= i - k:
                dq.popleft()

            while dq and dp[dq[-1]] <= dp[i]:
                dq.pop()

            dq.append(i)

        return dp[-1]

File: java/01696_jump_game_VI.java
// Time complexity - O (n)
// Space complexity - O (n)
public class Solution {
    public int maxResult(int[] nums, int k) {
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];

        Deque<Integer> deque = new LinkedList<>();
        deque.offer(0);

        for (int i = 1; i < n; i++) {
            dp[i] = nums[i] + dp[deque.peekFirst()];

            while (!deque.isEmpty() && deque.peekFirst() <= i - k) {
                deque.pollFirst();
            }

            while (!deque.isEmpty() && dp[deque.peekLast()] <= dp[i]) {
                deque.pollLast();
            }

            deque.offerLast(i);
        }

        return dp[n - 1];
    }
}
